daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/layers/api/Layer.java

```diff
@@ -27,51 +27,231 @@
+ *         this.direction = direction;
+ *         this.wordCount = wordCount;
+ *     }
+ * </pre>
+ * Beside initializing local members, this setter is also an initializer for the layer class instance.
+ * Any exception thrown becomes a Schema Definition Error.
+ * <p>
+ * If there are no parameter variables, then this setter, with no arguments, can be used purely for initialization.
  * <p>
- * A layer that wants to return a value after the layer algorithm completes defines a special recognizable
- * getter method. The name of the getter is formed from prefixing the DFDL variable name with the string
- * 'getLayerVariableResult_'. The return type of the getter must match the type of the variable.
+ * A DFDL variable used to return a result from a layer must be undefined, since variables in DFDL are single-assignment.
+ * Usually this means the use of the layer must be surrounded by a {@code dfdl:newVariableInstance}
+ * annotation which creates a new instance of the layer result variable, over a limited scope
+ * of use.
+ * The variable is assigned by the layer, and it is then available for reading by the DFDL schema until
+ * the end of the {@code dfdl:newVariableInstance} scope.
  * <p>
- * For example, a result value getter for a DFDL variable named 'checksum' of type xs:unsignedShort would be:
+ * To return a value into a DFDL variable, the layer implementation defines a special recognizable
+ * getter method.
+ * The name of the getter is formed from prefixing the DFDL variable name with the string
+ * "{@code getLayerVariableResult_}".
+ * The return type of the getter must correspond to the type of the variable.
+ * <p>
+ * For example, a result value getter for a DFDL variable named {@code total} of type {@code xs:unsignedShort} would be:
  * <pre>
- *      int getLayerVariableResult_checksum() {
- *          // returns the value created by the checksum algorithm.
+ * {@code
+ *      int getLayerVariableResult_total() {
+ *          // returns the value created by the layer's algorithm.
+ *          // commonly this returns the value of a data member.
+ *          return this.total;
  *      }
+ * }
  * </pre>
+ * Layers could have multiple result variables, but a single variable is most common, generally for returning checksums.
+ * See the {@link ChecksumLayer} class, which is designed to facilitate creation of checksum/CRC/hash/digest layers.
+ * <p>
+ * The Java types to use for the setter arguments and getter return types correspond to the DFDL variable types
+ * according to this table:
+ * <table border="1">
+ *     <tr>
+ *         <th>DFDL Schema Type</th>
+ *         <th>Java Type</th>
+ *     </tr>
+ *     <tr>
+ *         <td>xs:byte</td>
+ *         <td>byte</td>
+ *     </tr>
+ *     <tr>
+ *         <td>xs:short</td>
+ *         <td>short</td>
+ *     </tr>
+ *     <tr>
+ *         <td>xs:long</td>
+ *         <td>long</td>
+ *     </tr>
+ *     <tr>
+ *         <td>xs:int</td>
+ *         <td>java.math.BigInteger</td>
```

Comment:
Does xs:int map to a Java int? I thought xs:integer maps to BigInteger? And I think this list is missing xs:integer? Is that an accident or do layers not support xs:integer?

Reply:
Hmmm. I will check the code and/or fix this doc. xs:integer should have a mapping to java.lang.BigInteger and xs:int to int.